torchao/quantization/quantize_/workflows/float8/float8_tensor.py

```diff
@@ -0,0 +1,612 @@
+    """
+
+    tensor_data_names = ["qdata", "scale"]
+    tensor_attribute_names = [
+        "block_size",
+        "mm_config",
+        "hp_value_lb",
+        "hp_value_ub",
+        "act_quant_kwargs",
+        "kernel_preference",
+        "dtype",
+    ]
+
+    def __new__(
+        cls,
+        qdata,
+        scale,
+        block_size,
+        mm_config,
+        hp_value_lb,
+        hp_value_ub,
+        act_quant_kwargs,
+        kernel_preference,
+        dtype,
+    ):
+        shape = qdata.shape
+        kwargs = {}
+        kwargs["device"] = qdata.device
+        kwargs["dtype"] = dtype
+        kwargs["requires_grad"] = False
+        return torch.Tensor._make_wrapper_subclass(cls, shape, **kwargs)  # type: ignore[attr-defined]
+
+    def __init__(
+        self,
+        qdata: torch.Tensor,
+        scale: torch.Tensor,
+        block_size: Optional[List[int]] = None,
+        mm_config: Optional[Float8MMConfig] = None,
+        hp_value_lb: Optional[float] = None,
+        hp_value_ub: Optional[float] = None,
+        act_quant_kwargs: Optional[QuantizeTensorToFloat8Kwargs] = None,
+        kernel_preference: KernelPreference = KernelPreference.AUTO,
+        dtype: Optional[torch.dtype] = None,
+    ):
+        self.qdata = qdata
+        self.scale = scale
+        self.block_size = block_size
+        self.mm_config = mm_config
+        self.hp_value_lb = hp_value_lb
+        self.hp_value_ub = hp_value_ub
+        self.act_quant_kwargs = act_quant_kwargs
+        self.kernel_preference = kernel_preference
+
+    def __repr__(self):
+        return (
+            f"{self.__class__.__name__}({self.act_quant_kwargs=}, {self.qdata=}, {self.scale=}, "
+            f"{self.block_size=}, {self.mm_config=}, "
+            f"{self.shape=}, {self.device=}, {self.dtype=})"
+        )
+
+    def _quantization_type(self):
+        return f"{self.act_quant_kwargs=}, {self.block_size=}, {self.mm_config=}, {self.scale.shape=}"
+
+    def dequantize(self, output_dtype: Optional[torch.dtype] = None) -> torch.Tensor:
+        if output_dtype is None:
+            output_dtype = self.dtype
+
+        qdata, scale = self.qdata, self.scale
+        return _dequantize_affine_float8(qdata, scale, output_dtype)
+
+    @classmethod
+    def to_float8(
+        cls,
+        hp_tensor: torch.Tensor,
+        float8_dtype: torch.dtype = torch.float8_e4m3fn,
+        granularity: FP8Granularity = PerRow(),
+        mm_config: Optional[Float8MMConfig] = None,
+        hp_value_lb: Optional[float] = None,
+        hp_value_ub: Optional[float] = None,
+        kernel_preference: KernelPreference = KernelPreference.AUTO,
+        act_quant_kwargs: Optional[QuantizeTensorToFloat8Kwargs] = None,
+    ):
+        block_size = get_block_size(hp_tensor.shape, granularity)
+        block_size = list(block_size)
+
+        # for per row quantization and kernel_preference default setting, we'll use triton kernel for best performance
+        if (
```

Comment:
it can be in a separate PR, but I think we should test that all the `KernelPreference` choices are bitwise equivalent to each other, and ideally also test that the expected kernels are being called for each `KernelPreference` choice

Reply:
ok for separate PR, but I think this 100% should be done before public launch

Reply:
makes sense, will do in a follow up PR